dev/TeachingTip/TeachingTip.cpp

```diff
@@ -0,0 +1,1507 @@
+﻿#include "pch.h"
+#include "common.h"
+#include "TeachingTip.h"
+#include "RuntimeProfiler.h"
+#include "ResourceAccessor.h"
+#include "TeachingTipClosingEventArgs.h"
+#include "TeachingTipClosedEventArgs.h"
+#include "TeachingTipTestHooks.h"
+#include "TeachingTipAutomationPeer.h"
+
+TeachingTip::TeachingTip()
+{
+    __RP_Marker_ClassById(RuntimeProfiler::ProfId_TeachingTip);
+    SetDefaultStyleKey(this);
+    EnsureProperties();
+    SetValue(s_TemplateSettingsProperty, winrt::make<::TeachingTipTemplateSettings>());
+}
+
+winrt::AutomationPeer TeachingTip::OnCreateAutomationPeer()
+{
+    return winrt::make<TeachingTipAutomationPeer>(*this);
+}
+
+void TeachingTip::OnApplyTemplate()
+{
+    winrt::IControlProtected controlProtected{ *this };
+
+    m_beakOcclusionGrid.set(GetTemplateChildT<winrt::Grid>(s_beakOcclusionGridName, controlProtected));
+    m_contentRootGrid.set(GetTemplateChildT<winrt::Grid>(s_contentRootGridName, controlProtected));
+    m_nonBleedingContentRootGrid.set(GetTemplateChildT<winrt::Grid>(s_nonBleedingContentRootGridName, controlProtected));
+    m_bleedingImageContentPresenter.set(GetTemplateChildT<winrt::ContentPresenter>(s_bleedingImageContentPresenterName, controlProtected));
+    m_iconContentPresenter.set(GetTemplateChildT<winrt::ContentPresenter>(s_iconName, controlProtected));
+    m_actionButton.set(GetTemplateChildT<winrt::Button>(s_actionButtonName, controlProtected));
+    m_alternateCloseButton.set(GetTemplateChildT<winrt::Button>(s_alternateCloseButtonName, controlProtected));
+    m_closeButton.set(GetTemplateChildT<winrt::Button>(s_closeButtonName, controlProtected));
+    m_beakEdgeBorder.set(GetTemplateChildT<winrt::Grid>(s_beakEdgeBorderName, controlProtected));
+    m_beakPolygon.set(GetTemplateChildT<winrt::Polygon>(s_beakPolygonName, controlProtected));
+    if (SharedHelpers::IsThemeShadowAvailable())
+    {
+        m_shadowTarget.set(GetTemplateChildT<winrt::Grid>(s_shadowTargetName, controlProtected));
+    }
+
+    if (m_beakOcclusionGrid)
+    {
+        m_contentSizeChangedRevoker = m_beakOcclusionGrid.get().SizeChanged(winrt::auto_revoke, {
```

Comment:
NIT: unconditionally unhook all the event handler at the start of the function since it is possible for the template to change and for OnApplyTemplate to get called again.